[Woo POS] Fix for broken Home screen state after Activity death.

samiuelson opened the pull request:
<!-- Remember about a good descriptive title. -->
Fix for broken Home screen state after Activity death.

Closes: #12027

<!-- Id number of the GitHub issue this PR addresses. -->

### Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->
This PR fixes the way the state is recreated after Activity is recreated. 
* View models are now storing the state using SavedStateBundle.
* `HomeViewModel` ignores Products pane state events if Products pane is not currently visible (in the Checkout state). Previously it was wrongly modifying the Cart pane state.

### Steps to reproduce
<!-- Step-by-step testing instructions. For new user flows, consider instead stating the goal of the workflow and see if your PR reviewer can accomplish the workflow without specific steps! -->
On `trunk`:
1. Enable 'Don't keep activities' in dev settings
2. Open POS
3. Select a product
4. Click check out
5. Put the app into the background
6. Bring it back to the front
7. Notice "Checkout button" is missing and the cart has "back" arrow.

### Testing information
<!-- This is your opportunity to break out individual scenarios that need testing (when necessary) and/or include a checklist for the reviewer to go through. Consider documenting the following from your own completed testing: devices used, alternate workflows, edge cases, affected areas, critical flows, areas not tested, and any remaining unknowns. Provide feedback on this new section of the PR template through Sept 30, 2024 to Apps Quality; additional context here: p91TBi-b8z-p2#comment-12036 -->
Verify that the Home screen is in the correct state after returning back to the app.

### Images/gif
<!-- Include before and after images or gifs when appropriate. -->
|Before|After|
|---|---|
|<img src="https://github.com/user-attachments/assets/49cb43e9-54cc-465a-b90d-3b6c50d96c98">|![Screenshot_20240801_152148](https://github.com/user-attachments/assets/e3afc86f-6f37-4e25-acee-e40f01bf1561)|

- [x] I have considered if this change warrants release notes and have added them to `RELEASE-NOTES.txt` if necessary. Use the "[Internal]" label for non-user-facing changes.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

samiuelson committed:
Save and restore view state using SavedStateHandle

samiuelson committed:
Don't recompute checkout state when products state changes

dangermattic commented:
<!--
  0 Errors
  0 Warnings
  1 Message: This PR is still a Draft: some...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

samiuelson committed:
Fix detekt's complaints

samiuelson committed:
Merge branch 'trunk' into 12027-home-screen-broken-after-activity-death

samiuelson committed:
Update unit tests

samiuelson committed:
Merge remote-tracking branch 'origin/12027-home-screen-broken-after-activity-death' into 12027-home-screen-broken-after-activity-death

samiuelson committed:
Merge branch 'trunk' into 12027-home-screen-broken-after-activity-death

malinajirka commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/products/WooPosProductsViewModel.kt:
```diff
@@ -25,11 +27,17 @@ class WooPosProductsViewModel @Inject constructor(
     private var loadMoreProductsJob: Job? = null
 
-    private val _viewState = MutableStateFlow<WooPosProductsViewState>(WooPosProductsViewState.Loading())
+    private val _viewState: MutableStateFlow<WooPosProductsViewState> = savedStateHandle.getStateFlow(
```
⚠️ Saving complete ViewState into the Bundle can lead to TransactionTooLarge crashes, affect performance, and is in general discouraged ([more context here](https://github.com/woocommerce/woocommerce-android/pull/12105)). Have you considered recreating it after the process death instead of storing it?

samiuelson replied:
Good point @malinajirka, restoring of the products pane state is not necessary.

I reverted storing of the Products state in fe5b44c73f9ed44b4a781234e8246f0a64f4f7b1 and 80d918104d27d30918bfd21814b8973d404fb709.

We still store `WooPosCartState` in Bundle which can potentially exhaust capacity of Bundle though. I'd personally improve this (e.g. write the cart's state into DB) later in the next iterations. Let me know what you think!

malinajirka replied:
Thanks @samiuelson !

> We still store WooPosCartState in Bundle which can potentially exhaust capacity of Bundle though. I'd personally improve this (e.g. write the cart's state into DB) later in the next iterations. Let me know what you think!

I think we might want to consider storing only product-ids instead of storing the complete state. 

I'm fine if we improve it later on, however, I'm slightly worried these types of changes are often forgotten and overall lead to a lower quality of the codebase. 

Having said that, as long as we at least record an issue with a proposal on how to iterate on it, it's fine with me.

samiuelson replied:
I added the [issue](https://github.com/woocommerce/woocommerce-android/issues/12209) to the project board.

samiuelson committed:
Merge branch 'trunk' into 12027-home-screen-broken-after-activity-death

samiuelson committed:
Revert storing Products pane state in Bundle

samiuelson committed:
Revert parcelization of `WooPosProductsViewState`

malinajirka commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/WooPosHomeViewModel.kt:
```diff
@@ -129,6 +133,7 @@ class WooPosHomeViewModel @Inject constructor(
     }
 
     private fun handleProductsStatusChanged(event: ChildToParentEvent.ProductsStatusChanged) {
+        if (state.value.screenPositionState is WooPosHomeState.ScreenPositionState.Checkout) return
```
Tbh I'm not sure I follow this code, I'm wondering if comments or refactoring might help here. To me it seems, we are handling `Checkout.NotPaid` and `Checkout.Paid` states in the `when` condition below, but we are exiting the method just above it and ignoring these states. Wdyt?

What is also a bit confusing is that it seems we use `_state` and `state` interchangeably, is that intentional?

samiuelson replied:
I cleaned up the code in ffd17b6ee3f79c66b0e7d2351ea8e8bd014cfced. I believe the method is more readable now. It should result in the same effect though. 

In addition to that I added a test case (44b8c7934cf1b962f0c652726af55a99d41fb9ed) to document the intended behavior – when Products status is changed, in case Home screen is at the Checkout position, screen position state should not be modified. 

Let me know if you feel additional comment to the method is still nice to have.

malinajirka commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/products/WooPosProductsViewState.kt:
```diff
@@ -32,10 +30,9 @@ sealed class WooPosProductsViewState(
         WooPosProductsViewState(reloadingProductsWithPullToRefresh)
 }
 
-@Parcelize
```
Is this change intentional?

samiuelson replied:
Yes. AFAICS, implementing `Parcelable` by `WooPosProductsListItem` is redundant, because it's not serialized and stored into `Bundle` anywhere.

malinajirka reviewed (commented):
Thanks for working on this @samiuelson !

I've left a couple questions in the in-code comments. I'm also wondering whether you considered adding unit tests for this scenario - I'm not sure if it's feasible/worth it, but I was wondering if you gave it a though.

samiuelson committed:
Improve the code style

samiuelson committed:
Add test for activity death scenario

samiuelson committed:
Merge branch 'trunk' into 12027-home-screen-broken-after-activity-death

samiuelson commented:
> Thanks for working on this @samiuelson !
> 
> I've left a couple questions in the in-code comments. I'm also wondering whether you considered adding unit tests for this scenario - I'm not sure if it's feasible/worth it, but I was wondering if you gave it a though.

Thank you for review, @malinajirka 🙌🏼 

I addressed your comments. Since you reported the code is confusing I improved its style and added the test method for the sake of making its intent explicit. Let me know what you think.

samiuelson commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/WooPosHomeViewModel.kt:
```diff
@@ -129,17 +133,23 @@ class WooPosHomeViewModel @Inject constructor(
+            ChildToParentEvent.ProductsStatusChanged.FullScreen -> {
+                when (screenPosition) {
+                    is WooPosHomeState.ScreenPositionState.Cart -> WooPosHomeState.ScreenPositionState.Cart.Hidden
+                    is WooPosHomeState.ScreenPositionState.Checkout -> screenPosition
```
In case the Home screen is at the Checkout position, we don't want to change it in the event of the Products VM being recreated after activity death, emitting the `ChildToParentEvent.ProductsStatusChanged` event.

samiuelson committed:
Merge branch 'trunk' into 12027-home-screen-broken-after-activity-death

malinajirka commented:
Thanks for working on this @samiuelson and thank you for the changes! The code looks great to me! 🙇 


It seems it fixes the original issue, but most of the times when the activity gets recreated the `cart` is greyed out for some reason - clicking on the items still works, but it looks like if it was disabled - Emulator API 33 and OnePlus API 34.


![Screenshot_1723145480](https://github.com/user-attachments/assets/27daea71-fd8e-44ce-90c9-10f034b2d1e4)

samiuelson committed:
Update image visibility conditions

samiuelson committed:
Revert "Update image visibility conditions"

This reverts commit ed73579ba1a9840e14b637694daf9ad85aa934c6.

samiuelson committed:
Move responsibility of cart's opacity animating to cart

It's the Cart that keeps in its internal state the selected items count or empty state. That's why the Cart should decide about the opacity of its overlay and control the animation. Home is not keeping that state and is not capable of knowing if the Cart is really empty or not.

samiuelson commented:
> It seems it fixes the original issue, but most of the times when the activity gets recreated the `cart` is greyed out for some reason - clicking on the items still works, but it looks like if it was disabled - Emulator API 33 and OnePlus API 34.

Thanks, @malinajirka for a thorough check—good catch 🙇🏼  I believe the best solution is to move control over the Cart overlay's opacity from Home to Cart because Home doesn't really know if the Cart is empty or not. Fixed in this commit - 3297e5c

Let me know what you think!

samiuelson committed:
Remove unused imports

malinajirka reviewed (approved):
LGTM, thanks @samiuelson !
